### ISSUE TITLE
RFC 0056 & RFC 0023 Usage of inline public keys

### ISSUE THREAD

**Issue by tplooker** [hyperledger/aries-rfcs#201]
A reoccurring issue has been how to sufficiently represent inline keys in a concise syntax whilst preserving the required information around a public keys encoding and underlying type.

The [did:key](https://github.com/digitalbazaar/did-method-key) method appears to solve these issues by leveraging the following list of [multicodecs](https://github.com/multiformats/multicodec/blob/master/table.csv).

Using this method we could replace any current references we currently have to inline public keys that at present do not include any information about the public keys encoding or type.

An example of this in the service decorator instance would be the following.

```
{
    "@type": "somemessagetype",
    "~service": {
        "recipientKeys": ["B12NYF8RrR3h41TDCTJojY59usg3mbtbjnFs7Eud1Y6u"],
        "routingKeys": ["B12NYF8RrR3h41TDCTJojY59usg3mbtbjnFs7Eud1Y6u"]
        "serviceEndpoint": "https://example.com/endpoint"
    }
}
```

Would change to 

```
{
    "@type": "somemessagetype",
    "~service": {
        "recipientKeys": ["did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH"],
        "routingKeys": ["did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH"]
        "serviceEndpoint": "https://example.com/endpoint"
    }
}
```

Whereby resolving`did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH` would yield the following 

```
{
  "@context": "https://w3id.org/did/v1",
  "id": "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH",
  "publicKey": [
    {
      "id": "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH",
      "type": "Ed25519VerificationKey2018",
      "controller": "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH",
      "publicKeyBase58": "B12NYF8RrR3h41TDCTJojY59usg3mbtbjnFs7Eud1Y6u"
    }
  ],
  "authentication": [
    "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH"
  ],
  "assertionMethod": [
    "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH"
  ],
  "capabilityDelegation": [
    "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH"
  ],
  "capabilityInvocation": [
    "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH"
  ],
  "keyAgreement": [
    {
      "id": "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH#zBzoR5sqFgi6q3iFia8JPNfENCpi7RNSTKF7XNXX96SBY4",
      "type": "X25519KeyAgreementKey2019",
      "controller": "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH",
      "publicKeyBase58": "JhNWeSVLMYccCk7iopQW4guaSJTojqpMEELgSLhKwRr"
    }
  ]
}
```

**Comment by tplooker** [comment]
@dhh1128 @kdenhartog @TelegramSam

**Comment by dhh1128** [comment]
For the time being, I am STRONGLY opposed to introducing a dependency on did:key. I don't agree with any of the extra semantics that this method wants to introduce -- specifically the ones related to `assertionMethod`, `capabilityDelegation`, `capabilityInvocation`, and `keyAgreement`. This is because, while I love the OCAP model, the DB method is linked to an implementation of OCAP (OCAP-LD) which I consider dangerous and unnecessary; Sovrin-style credentials are a safer, cheaper, and more privacy-respecting OCAP solution because of the way they test the delegation chain and revocation properties. Now, I am not trying to argue the OCAP issue in this ticket, and it's entirely possible that we could converge over time--but I don't want to jump on the bandwagon of this DID method and accept a bunch of assumptions that we later have to unwind.

However, I think the key encoding mechanism is clever and useful. So why don't we do what you are proposing, except not use the "did:key:" prefix. That is, just provide the keys in encoded form, such that if we decide to use did:key later, we can--and if we don't, we haven't lost anything because we still have the property we truly need.

**Comment by swcurran** [comment]
I agree with Daniel. The encoding is very useful, but the highly opinionated derived DIDDoc is asking for trouble. If there was a way to define our own DIDDoc template, that would be helpful, but that not be `did:key`.  I think that without explaining to developers the purpose of all the fields in the generated DIDDoc (using phrases other than "ignore this"), we're asking for trouble.

I do like the concise encoding.

**Comment by tplooker** [comment]
@dhh1128 point taken, perhaps more education and understanding of some of the implications that this method entails is required. Are you prepared to document some of your concerns so we can evaluate, alternatively I'm happy to jump on a call to discuss further, or perhaps we can put this on the agenda for the WG call? To me inventing a new syntax that has no prefix feels like wasting/duplicating alot of the work of the DID spec which is to establish a uniform data model behind these style of identifiers.

**Comment by tplooker** [comment]
Tagging #104 explicitly because of its overlap I didn't see until I opened this issue

**Comment by tplooker** [comment]
Thanks, I do understand that the multi-codec identifier becomes the prefix, however I'm still not convinced with this approach. I understand that `did:key` is still forming and hence is not a complete solution but I am wondering if this is at least an opportunity to collaborate on forming a generalized ephemeral did solution.

The core of my argument is what do we want to refer to other parties in the ecosystem as, keys or some more generalized type of identifier? Because supporting both seems like a strange fork. The solution in my mind is simple whenever you are talking about a key that belongs to a did doc, reference it via a did-url, because chances are if you are authenticating a message from or sending a message to, you will be speaking about them in the context of their did not key. If it is an ephemeral key you are authenticating a message from or sending a message to, why not construct an ephemeral did so that the way in which you identify subjects is uniform. Otherwise in parsing fields like `recipientKeys` or `routingKeys` you will have to always have a branch of logic that is looking for the did prefix passing it to a did resolver and another branch of logic that tests if its multi-codec public key and instead perform some piece wise logic that is non-standard and will return
some implementation specific understanding of the public key. This duplication creates headaches wherever this functionality is required and I see little benefit to effectively creating another non-standard identifier to a public key.

Another important observation to note is although DID's were not designed to be entirely human memorable, they are in most cases human identifiable, which is an important distinction to make. Because if we invest in creating a multi-codec form with no human readable prefix like `did:key` this will potentially be the source of much confusion, i.e how do I as a human disambiguate between my base64 public key that was spat out of some system versus my multi-codec public key, see an extract taken from the example refereed to above.

`B12NYF8RrR3h41TDCTJojY59usg3mbtbjnFs7Eud1Y6u` => `z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH`

Or with did:key prefix it becomes far less ambiguous.

`B12NYF8RrR3h41TDCTJojY59usg3mbtbjnFs7Eud1Y6u` => `did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH`

In regards to your second and third concerns raise in your points above, my understanding of the applications of OCAP's is different to what you describe, in particular how you describe validating attenuated delegation in an OCAP model. For instance, the delegation chain can be embedded into an OCAP rather than having to traverse a chain and the revocation check can be either be a call back to the original issuer or it could reference some global revocation infrastructure like what Sovrin provides. The point i'm trying to make is that is an implementation specific detail and I don't think did:key (but I may well be wrong) has hard opinion on the exact implementation and even in the event it does, is there still an opportunity to discuss this as an issue with one possible usage of did:key rather than something fundamental to did:key?

**Comment by tplooker** [comment]
@dhh1128 to your point about SSH above, unless you are talking about putting raw public keys in the fields described which re-introduces the ambiguity we are trying to solve, you will still have to cast them to some other representation before using i.e a multi-codec base or something like `did:key`

**Comment by tplooker** [comment]
@SmithSamuelM I am happy to explore a more pure ephemeral did too if we deem did:key is too opinionated about its usage, what I am opposed to is creating a new identifier class i.e just using the multi-codec prefix to a public key.

**Comment by tplooker** [comment]
@SmithSamuelM correct in my mind with an ephemeral DID you would use a DID resolver that would just inflate the DID into a minimal DID Doc quoting one key, so no actually resolution required. Conceptually ephemeral DIDs are similar to did:peer in the fact they do not rely on a ledger, but they go one step further and remove the dependency on local state and maintaining local state. It is important to note I'm not raising this comparison to justify ephemeral DID's as a replacement to peer DID's, I think there are very valid use-cases for both, I'm merely pointing out that we already have an example of a did not anchored to a ledger.

**Comment by tplooker** [comment]
Also creating another identifier standard is not a small undertaking, so we should approach this from the angle of why ephemeral did's do or do not fit under the did spec before we decide to create a new type of identifier, rather than the other way around.

**Comment by SmithSamuelM** [comment]
The core feature of a self certifying identifier is that the public key of a public private signing key pair is included in the identifier (or at least a fingerprint).  These means that any attestations signed with the private key may be verified with the public key in the identifier.  If all one needs is a public private key pair to control the attributes associated with an identifier then the identifier provides the public key.   The assumption is that the crypto suite used for the associated signature is known. If the crypto suite is also provided in the identifier via query parameters for example then that assumption goes away and the tooling is now future proofed. Control over the attributes is proven via signatures and the identifier provides alll the information a verifier needs to verify.  Notice that in this case no resolution aka no DDo is needed for this simple case.  The primary purpose of a DDo is to support a persistent did that must survive for multiple uses. One of the things that happens is keys are exploited due to exposure so the did doc provides a way to rotate the private key but still keep the same public key embedded in the identifier. In other words the DDo enables key management for the long term viability of a persistent did.  A secondary purpose of a did document is indirection to services affiliated with a given did.  An ephemeral did does not have long term key management or affiliated services so has no need for a did document. It’s a one time use key pair or short time use.
However tooling for DIDs may involve resolution so an ephemeral did is resolvable but trivially so.

The purpose of the did method is to provide authenticated CRUD operations for updating the two purposes 1) key management. 2) indirection to affiliated services. Anything more than this will overly complicating did resolvers. Any new features besides key management may be provided with the indirected services. Because an ephemeral did is not persistent the did document may never need update for rotating keys but merely crypto suite updates. 

For encryption keys a derivation algorithm may be used from the private key and a path in the query string. This also may be used to bootstrap a Diffie Hellman exchange.  Nowhere is did resolution required because the did itself is self contained as long as the crypto suite is known. Additional properties like detecting stale ephemeral DIDs might be useful. In this case did resolution may or may not have a use.

Did resolution is all about key management and indirection but most important is key management.

**Comment by tplooker** [comment]
@dhh1128 to be clear are you saying the `recipientKeys` and `routingKeys` fields should only ever be populated with this new type of concise identifier, which is a multi-codec public key?

**Comment by peacekeeper** [comment]
I'm a bit late to a long discussion, but I like the following points made by @dhh1128:

> the items we're listing in ~service.recipientKeys and ~service.routingKeys are keys, not DIDs. Putting DIDs where keys belong is an error

> what we are listing in the recipientKeys and routingKeys arrays are really, truly keys. They are not identifiers for agents.

The other issue https://github.com/hyperledger/aries-rfcs/issues/104 has good thoughts on whether keys should be passed as inline values or as references, but in this issue here the original question seems to be whether inline keys should be passed as bare multicodec'ed values or as `did:key`. I think this decision should depend on whether something is being identified or not - as @dhh1128 put it in [4) We're not identifying](https://github.com/hyperledger/aries-rfcs/issues/104#issuecomment-525396519).

Whenever a DID is used somewhere, let's ask the question _"What's the DID subject?"_. If there's no good answer to that question, then probably just a (multicodec'ed) key value should be used instead of a `did:key`.

I suspect the answer to that question is linked to the discussion in the other thread on how much separation or integration there should be between the crypto layer and the DID layer.

**Comment by peacekeeper** [comment]
@SmithSamuelM

> Because an ephemeral did is not persistent

This is just a minor detail, but I would argue a DID that doesn't support the Update or Deactivate operations is still "persistent", i.e. it will never be re-assigned to identify a different subject.

> I think we should have an ephemeral did not did:key

I'm not sure I understand why you would not consider `did:key` to be "ephemeral" or what exactly you mean by that?

**Comment by swcurran** [comment]
Discussed on the Aries Working Group call 2019.11.27.  We tried to see if we could close this issue with a recommendation, but that didn't happen.  Relevant points:

- Counter to my earlier [comment](https://github.com/hyperledger/aries-rfcs/issues/201#issuecomment-524476612) that many of the entries in the DIDDoc would not be useful in DIDComm, Tobias pointed out that likely most would be useful, including "keyAgreement" where a Curve25519 ECDH key is derived from an ed25519 public key.
- Tobias pointed out that if we don't do this, we have to document somewhere else, in a form similar to did:key how another party is expected to use the key. If we use did:key, that documentation and understanding will have been done.
- Not doing at least something - multicodecs or did:key seemed like a bad approach.

We hope to talk about this at the Connectathon next week and move this to the point that we can both take action on this and close this issue.

**Comment by llorllale** [comment]
@swcurran @kdenhartog 

> * Counter to my earlier comment that many of the entries in the DIDDoc would not be useful in DIDComm, Tobias pointed out that likely most would be useful, including "keyAgreement" where a Curve25519 ECDH key is derived from an ed25519 public key.

I would like to know more about why some existing implementations are deriving the encryption key from a signing key. As far as I am aware, [DID docs can include keys for "digital signatures, encryption and other cryptographic operations"](https://w3c.github.io/did-core/#public-keys), so I don't see the reason for this conversion.

**Comment by tplooker** [comment]
@llorllale, I'm not sure I fully understand you question? Are you asking why are we using a single key for both encryption and signing?

**Comment by llorllale** [comment]
@tplooker not quite; I'm asking why some implementations derive the encryption key from the signing key (the RFC says the input is a signing key), instead of starting with an encryption key as input in the first place.

**Comment by swcurran** [comment]
@llorllale - that specific implementation's do that (or not) is a separate issue from this discussion. A discussion about that with the implementers would be needed. Perhaps that is a discussion needed around the did:peer DIDDocs that we expect - I'm not sure.

In the case being discussed here, we have (by definition) just one public key (a string) + the algorithm type. The transformation that is done with `did:key` is to take those two values and generate a standard DIDDoc with (mostly) string replacement, but in the case of the "keyAgreement", the derivation of a new public key from the initial one.

**Comment by SmithSamuelM** [comment]
One very good reason to start with a signing public/private key pair and then derive an encryption public/private key pair from the signing key pair is that now only one private key must be stored / remembered recovered.  This reduces the key management load are risk on the user which IMHO is the gating complexity for broad adoption of decentralized identifiers.  Combining this approach of deriving encryption keys from signing keys with an HDKey algorithm where any number of signing key pairs may be re-derived from a single private key plus a public path per key then one starts to get scalable key management.  So the HD keychain for signing becomes for free also an HD Key chain for encryption.

**Comment by Baha-sk** [comment]
I disagree with deriving/converting encryption keys from signature keys as this poses a security threat. If a key is compromised, the hacker can derive/convert other keys and do more harm with only 1 key. This is more true for Ed25519 to X25519 conversions as they're 1 to 1 conversions. If a hacker lays their hands on the signature key, they can decrypt all DIDComm messages. To avoid this, signature keys should not be linked to encryption keys in any way.

**Comment by llorllale** [comment]
Note: this has implications for the out-of-band, did-exchange and connections protocols as they exist today, where the `response` is signed with a `recipientKey` sent in the `invitation`

**Comment by swcurran** [comment]
So discussion for tomorrow at the Aries WG call?  @Baha-sk @llorllale -- can you lead the discussion?  Your two notes cover it nicely - whether or not it is a safe practice, and if not, how it would affect what we are doing in various protocols.

Please confirm if you can make it -- I'm prepping the agenda now.

**Comment by SmithSamuelM** [comment]
@Baha-sk   "I disagree with deriving/converting encryption keys from signature keys ". I agree with you in the general case. The one case, however, where this is acceptable in my view is to bootstrap ephemeral communications which is the peer did case.  The exploit you talk about "lay their hands on the signature key" is just as valid an exploit for encryption keys as it is for signature keys. If your key management system is sufficiently weak that someone can "lay their hands" on your signature keys, what stops them from also laying their hands on your encryption keys as well?  If on the other hand you adequately protect your signature keys, then deriving an encryption key from a signature key is only a little less hard to brute force exploit (as it its very very hard).  The root of the problem is how do you manage your secrets.  If your management of signature keys is so bad that you must protect your encryption keys using some other mechanism then your most serious problem is not that you are converting signing keys to encryption keys but how you are protecting your signing keys in the first place.    There are some who erroneously hold that protecting confidentiality is more important than protecting authentication. I would argue that confidential but unauthenticated  communication is much less useful than authenticated communication (confidential or not). There are some edge cases where you want confidential but unauthenticated communication but for real transactions it must be authenticated first and always. So protecting your signing keys is a first priority.   Obfuscation of encryption keys is not protection. So protect your secrets well. Derived private keys from well protected secrets is the most practical approach to scalable key management.  Not having derivable keys or a key management infrastructure based on derived keys will not scale and makes decentralized identity impractical.

**Comment by SmithSamuelM** [comment]
If you are going to encrypt lots of data (not merely bootstrap communications). Then you wouldn't want to use only one encryption key (derived from a signing key or not) you would want to use rotated encryption keys at least one per session and maybe one per so many MB of data. this gives you forward secrecy. So even if the first encryption key is derived from a signing key (for the bootstrap). The subsequent rotated (or ratcheted) encryption keys are not exposed should the original signing key be compromised.   Simplistic rules of thumb about key management are usually wrong. Real world means you must rotate keys and build in support for key rotation. So if DID Comm is only bootstrapping a short set of messages one time with a set of keys, its not a problem. If you are using the same did comm keys indefinitely then that will eventually be a problem.

**Comment by SmithSamuelM** [comment]
In the general case for a key management approach based on derived keys, one starts with a root secret of sufficient entropy. This root must be extremely well protected. Then keys are derived from this root using a derivation process that preserves the cryptographic strength of the root. The derivation process should preserve the degree of cryptographic strength as a design parameter for your infra-structure (128 bits) is considered sufficient pre-quantum. Brute force exploit of 128 bits of cryptographic strength is a practical impossibility (pre-quantum). These first order derived keys may be roots in and of themselves but inhabit different infrastructure from the master root. This is a multi-valent key management infrastructure.  Then other keys within the same infrastructure may be derived from these 1st order roots. These are the keys that actually get used or exposed. With self-certifying identifiers the public keys become part of the identifier are are used to derive the identifier.  All the derived identifiers may be either ephemeral (one time short time use) or persistent. Persistent identifiers must be rotatable (ie their keys must be rotatable).   Given this approach, all exploits become key management exploits that are recoverable up to but not including a successful compromise of the master root secret.    When using multi-sig across entity boundaries then there may be multiple master secrets.    So the critical analysis for any did com  identifier is  to answer the question. Is its use ephemeral or persistent?  If ephemeral make sure it really is ephemeral (one time short time). If persistent then it must be rotatable.

**Comment by SmithSamuelM** [comment]
If is truly ephemeral then the derivation of an encryption key from a signing key is not significant security risk. If persistent then the signing key must be rotatable and rotated in which case so is the encryption key or the encryption keys is separably rotatable.   But using non-rotatable keys in a persistent application is a no-no no matter whether you are deriving encryption keys from signing keys or not.  Its the wrong question.

**Comment by SmithSamuelM** [comment]
The reason the DNS/CA is insecure is because convenient key rotation was so important that not having a secure root-of-trust for each identifier was preferable to more difficult key rotation from a secure root-of-trust.  Now that we know better, i.e. trusted entities are never trustable, we need secure roots-of-trust and we need to solve the key rotation problem from secure roots-of-trust. Anything less is a half measure and we have not made material progress in decentralized identity.  Hence the basis for the design of KERI.

**Comment by Baha-sk** [comment]
Thanks @SmithSamuelM for the detailed explanation. It sounds like you have described an advanced KMS system in a few comments. I'll definitely read about KERI. I found this article online: https://arxiv.org/pdf/1907.02143.pdf. Nice to meet you :-)

@swcurran the proposal is simple for DID exchange: add a second list of keys, this way we have one list dedicated to encryption only and another for verification. We can discuss during the Aries WG tomorrow, @llorllale and I will try to make it. Are you referring to the morning or the afternoon call?

**Comment by swcurran** [comment]
It's the afternoon call tomorrow - 3PM Eastern.  Thanks for joining -- much appreciated.
